### DISCUSSION TITLE
How can I tell whether the server is ready and open a web browser then? How do I find out if the server couldn't start because the port was busy?

### DISCUSSION THREAD

**Discussion by ilyagr** [poem-web/poem#751] [Q&A] [answered]
I would like to start a local Poem server and, when it's ready, open a webbrowser with something along the lines of the [`open` crate](https://docs.rs/open/latest/open/).

Currently, I have 
```rust
  let app: Route = /* ... */;
  Server::new(TcpListener::bind("127.0.0.1:8080")).run(app).await?;
```

How do I schedule something like `tokio::task::spawn_blocking` to start when the server is ready to receive connections?

Also, if the port I'm listening too is busy, how do I try other ports? Is this what `combined_listeners` example does, or is it listening on a bunch of ports?

**Update:** For the second question, I realized that `run(app).await` returns an error `Err(Os { kind: AddrInUse })` when the port is busy. Is checking for this the best way to go, or is this OS-dependent?

**Comment by ilyagr** [comment, marked as answer]
I have a theory of how to do this, based on the difference between an "Acceptor" and a "Listener". I'm not sure whether it's correct. I'll experiment if I don't hear from anybody.

Here's the outline I'm thinking about.

```rust
let mut port = 8000;
let acceptor = loop {
  if port > max_port {
    panic!();
  }
  // This seems like it never fails, unlike
  // https://doc.rust-lang.org/nightly/std/net/struct.TcpListener.html#method.bind
  let listener = TcpListener::bind(format!("127.0.0.1:{port}");
  match listener.into_acceptor() {
     Some(a) => return a,
     Err(_) => { port += 1} // Could also check the ErrorKind of this error, perhaps?
  }
}

println!("Server is ready for incoming connections at this point!");
println!("I'm hoping that, at this point, a browser connecting to \
         127.0.0.1:{port} will wait for the server to start and process the request.");
start_web_browser();
Server::new(acceptor).run(app).await?;
```


**Update:** https://docs.rs/poem/latest/src/poem/listener/tcp.rs.html#30-37 seems to confirm my theory.

**Reply by sunli829** [reply]
you are right

**Reply by sunli829** [reply]
When you convert a `Listener` to `Acceptor`, it means that the port has been bound, and you need to use `Server::new_with_acceptor` to create the server.

https://docs.rs/poem/latest/poem/struct.Server.html#method.new_with_acceptor

**Reply by ilyagr** [reply]
Thank you! Yeah, it seems to work.

Perhaps this should be an official example.

**Reply by ilyagr** [reply]
I made an example: https://github.com/poem-web/poem/pull/754
